### PR TITLE
AP_Networking: auto-restart PPP on error

### PR DESCRIPTION
**Pull request by tridge**
this makes it easier to handle startup timing with PPP

### CODE REVIEW

**Comment by timtuxworth** [comment]
Tested and working.

Test case: Durandal with Raspberry Pi Zero, installed in a plane with. a single battery powering the Durandal via a PDB and the Pi via a BEC.
pppd is installed on the Pi as a service, so it starts automatically when the Pi boots. Before this fix, it would fail every time unless the Durandal was manually rebooted from Mission Planner while the Pi was attempting to connect.

Now ppp comes up automagically and the web server is live a few seconds after the Pi finishes booting.

File: libraries/AP_Networking/AP_Networking_PPP.cpp [hunk 1]
```diff
@@ -169,28 +174,33 @@ void AP_Networking_PPP::ppp_loop(void)
...
+            auto n = uart->read(buf, sizeof(buf));
+            if (n > 0) {
+                LWIP_TCPIP_LOCK();
+                pppos_input(ppp, buf, n);
+                LWIP_TCPIP_UNLOCK();
+            } else {
+                hal.scheduler->delay_microseconds(200);
+            }
         }
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO, "PPP: restarting");
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking_PPP.cpp:203]: Is this GCS msg necessary?

File: libraries/AP_Networking/AP_Networking_PPP.h [hunk 2]
```diff
@@ -21,6 +21,7 @@ class AP_Networking_PPP : public AP_Networking_Backend
     AP_HAL::UARTDriver *uart;
     struct netif *pppif;
     struct ppp_pcb_s *ppp;
+    bool need_restart = false;
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking_PPP.h:24]: Remove the assignment

**Review by magicrub — CHANGES_REQUESTED** [review]

File: libraries/AP_Networking/AP_Networking_PPP.cpp (hunk 1 above)
Comment (timtuxworth) [reply libraries/AP_Networking/AP_Networking_PPP.cpp:203]: I agree - when I saw it in the messages, it didn't help me with anything at all, however just before this you see something like PPP timeout - which also seems unnecessary.

File: libraries/AP_Networking/AP_Networking_PPP.cpp [hunk 3]
```diff
@@ -71,6 +71,7 @@ void AP_Networking_PPP::ppp_status_callback(struct ppp_pcb_s *pcb, int code, voi
     }
     case PPPERR_OPEN: {            /* Unable to open PPP session. */
         msg = "PPPERR_OPEN";
+        driver.need_restart = true;
```
Comment (timtuxworth) [inline libraries/AP_Networking/AP_Networking_PPP.cpp:74]: The message should be something like "PPP: Unable to open session"

File: libraries/AP_Networking/AP_Networking_PPP.cpp [hunk 4]
```diff
@@ -87,6 +88,7 @@ void AP_Networking_PPP::ppp_status_callback(struct ppp_pcb_s *pcb, int code, voi
     }
     case PPPERR_CONNECT: {         /* Connection lost. */
         msg = "PPPERR_CONNECT";
+        driver.need_restart = true;
```
Comment (timtuxworth) [inline libraries/AP_Networking/AP_Networking_PPP.cpp:91]: The message should be something like "PPP: Connection Lost"

File: libraries/AP_Networking/AP_Networking_PPP.cpp [hunk 5]
```diff
@@ -99,14 +101,17 @@ void AP_Networking_PPP::ppp_status_callback(struct ppp_pcb_s *pcb, int code, voi
     }
     case PPPERR_PEERDEAD: {        /* Connection timeout */
         msg = "PPPERR_PEERDEAD";
+        driver.need_restart = true;
         break;
     }
     case PPPERR_IDLETIMEOUT: {     /* Idle Timeout */
         msg = "PPPERR_IDLETIMEOUT";
+        driver.need_restart = true;
```
Comment (timtuxworth) [inline libraries/AP_Networking/AP_Networking_PPP.cpp:109]: There should be no message for this one I think.

File: libraries/AP_Networking/AP_Networking_PPP.cpp [hunk 6]
```diff
@@ -99,14 +101,17 @@ void AP_Networking_PPP::ppp_status_callback(struct ppp_pcb_s *pcb, int code, voi
...
         break;
     }
     case PPPERR_CONNECTTIME: {     /* Max connect time reached */
         msg = "PPPERR_CONNECTTIME";
+        driver.need_restart = true;
```
Comment (timtuxworth) [inline libraries/AP_Networking/AP_Networking_PPP.cpp:114]: There should be no message for this one I think.

**Review by timtuxworth — CHANGES_REQUESTED** [review]
Since I started testing this PR, I noticed that the messages are not very helpful, so I'm making my comments here, although it doesn't strictly belong here:

        **msg = "PPPERR_IDLETIMEOUT";
        perhaps no message required for this one - it will just restart and keep trying

        msg = "PPPERR_CONNECTTIME";
        perhaps no message required for this one - it will just restart and keep trying**

        msg = "PPPERR_PARAM"; 
        perhaps better "PPP: Invalid parameter"

        msg = "PPPERR_OPEN";
        perhaps better  "PPP: unable to open session"

        msg = "PPPERR_DEVICE";
        perhaps better "PPP: Invalid I/O device"

        msg = "PPPERR_ALLOC";
        perhaps better "PPP: Unable to allocate resources"

        msg = "PPPERR_USER";
        perhaps better "PPP: User interrupt"

        msg = "PPPERR_CONNECT";
        perhaps better "PPP: Connection Lost"

        msg = "PPPERR_AUTHFAIL";
        perhaps better "PPP: authentication fail"

        msg = "PPPERR_PROTOCOL";
        break;

        msg = "PPPERR_PEERDEAD";
        perhaps better "PPP: connection lost"

        msg = "PPPERR_LOOPBACK";
        perhaps better "PPP: Loopback detected"

File: libraries/AP_Networking/AP_Networking_PPP.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking_PPP.cpp:203]: I agree about changing the other messages too. Error messages, sure, but normal behavior shouldn't need to spam the gcs

File: libraries/AP_Networking/AP_Networking_PPP.cpp [hunk 7]
```diff
@@ -169,28 +174,33 @@ void AP_Networking_PPP::ppp_loop(void)
...
+
+        while (!need_restart) {
+            uint8_t buf[1024];
+            auto n = uart->read(buf, sizeof(buf));
+            if (n > 0) {
+                LWIP_TCPIP_LOCK();
+                pppos_input(ppp, buf, n);
+                LWIP_TCPIP_UNLOCK();
+            } else {
+                hal.scheduler->delay_microseconds(200);
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking_PPP.cpp:200]: 200us seems pretty short for this sort of thing.

File: libraries/AP_Networking/AP_Networking_PPP.cpp [hunk 8]
```diff
@@ -169,28 +174,33 @@ void AP_Networking_PPP::ppp_loop(void)
...
+        LWIP_TCPIP_LOCK();
+        ppp_connect(ppp, 0);
+
+        netif_set_default(pppif);
+        LWIP_TCPIP_UNLOCK();
+
+        need_restart = false;
+
+        while (!need_restart) {
+            uint8_t buf[1024];
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking_PPP.cpp:193]: Curious, does moving this outside of the loop make any difference?

**Review by magicrub — CHANGES_REQUESTED** [review]

File: libraries/AP_Networking/AP_Networking_PPP.cpp (hunk 7 above)
Comment (tridge) [reply libraries/AP_Networking/AP_Networking_PPP.cpp:200]: that is existing code. The delay is for between PPP packets so we don't hog the CPU, 200us is plenty

File: libraries/AP_Networking/AP_Networking_PPP.cpp (hunk 8 above)
Comment (tridge) [reply libraries/AP_Networking/AP_Networking_PPP.cpp:193]: not really, it's a stack allocation, doesn't matter a lot where we move the stack pointer

File: libraries/AP_Networking/AP_Networking_PPP.cpp (hunk 6 above)
Comment (tridge) [reply libraries/AP_Networking/AP_Networking_PPP.cpp:71]: i've changed the unexpected ones to "PPP: code NN"

File: libraries/AP_Networking/AP_Networking_PPP.h (hunk 2 above)
Comment (tridge) [reply libraries/AP_Networking/AP_Networking_PPP.h]: done

File: libraries/AP_Networking/AP_Networking_PPP.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_Networking/AP_Networking_PPP.cpp]: I've changed it to "PPP: connected"
none of this is "normal behaviour"

File: libraries/AP_Networking/AP_Networking_PPP.cpp (hunk 6 above)
Comment (timtuxworth) [reply libraries/AP_Networking/AP_Networking_PPP.cpp:71]: That's much cleaner. Thanks Tridge.

**Review by magicrub — APPROVED** [review]
Looks good!

**Pull request merged**